UnitySDK/Assets/ML-Agents/Scripts/Agent.cs

```diff
@@ -288,9 +274,33 @@ void OnDisable()
                 Academy.Instance.AgentAct -= AgentStep;
                 Academy.Instance.AgentForceReset -= _AgentReset;
             }
+            OnDisableHelper();
             m_Brain?.Dispose();
         }
 
+        void OnDisableHelper()
+        {
+            if (m_VectorSensorBuffer == null)
+            {
+                // Create a buffer for writing uncompressed (i.e. float) sensor data to
+                m_VectorSensorBuffer = new float[sensors.GetSensorFloatObservationSize()];
+            }
+
+            // This is a bit of a hack - if we're in inference mode, observations won't be generated
+            // But we need these to be generated for the recorder. So generate them here.
+            var observations = new List<Observation>();
+            GenerateSensorData(sensors, m_VectorSensorBuffer, m_WriteAdapter, observations);
```

Comment:
I'm confused - you said before that the problem was some of the agent sensors might not be availble for generation (e.g. camera already cleaned up). If that's the case, won't that still be a problem here? How's this different from passing the actual sensors to RequestDecision?

Reply:
We generate the data earlier. OnDisable is called first on all destroyed gameObjects and then On Destroy is called on all game objects. 
We call GenerateSensorData in OnDisable rather than in the next fixed update.

Reply:
I'm generally confused about what this function is attempting to achieve.